src/main/java/frc/robot/commands/Shoot.java

```diff
@@ -0,0 +1,279 @@
+package frc.robot.commands;
+
+import edu.wpi.first.math.geometry.Pose3d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Rotation3d;
+import edu.wpi.first.wpilibj.DriverStation;
+import edu.wpi.first.wpilibj.Timer;
+import edu.wpi.first.wpilibj.DriverStation.Alliance;
+import edu.wpi.first.wpilibj2.command.Command;
+import frc.robot.constants.ArmConstants;
+import frc.robot.constants.ShooterConstants;
+import frc.robot.constants.StorageIndexConstants;
+import frc.robot.constants.miscConstants.VisionConstants;
+import frc.robot.subsystems.Drivetrain;
+import frc.robot.subsystems.gpm.Arm;
+import frc.robot.subsystems.gpm.Shooter;
+import frc.robot.subsystems.gpm.StorageIndex;
+
+//00 is the bottom right corner of blue wall in m
+/**
+ * Shoots on the move (instantaneous velocities and pose ver.).
+ */
+
+// TODO make the version with delay have correct math
+public class Shoot extends Command {
+        private final Shooter shooter;
+        public final Arm arm;
+        public final Drivetrain drive;
+        private final StorageIndex index;
+
+        private final Timer shootTimer = new Timer();
+
+        // for testing sakes
+        public double horiz_angle;
+        public double vert_angle;
+        public double exit_vel;
+        public Pose3d displacement;
+        public double v_rx;
+        public double v_ry;
+        // TODO: put in constants for other commands to use
+        private final double REST_VEL = 4; // TODO: determine the fastest idle note-exit velocity that won't kill the
+                                           // battery.
+
+        public Shoot(Shooter shooter, Arm arm, Drivetrain drivetrain, StorageIndex index) {
+                this.shooter = shooter;
+                this.arm = arm;
+                this.drive = drivetrain;
+                this.index = index;
+                addRequirements(shooter, arm, index);
+        }
+
+        @Override
+        public void initialize() {
+                // Reset the timer
+                shootTimer.reset();
+                shootTimer.stop();
+                drive.setIsAlign(true); // Enable alignment mode on the drivetrain
+        }
+
+        @Override
+        public void execute() {
+                // Positive x displacement means we are to the left of the speaker
+                // Positive y displacement means we are below the speaker.
+                Pose3d speakerPose = DriverStation.getAlliance().isPresent() &&
```

Comment:
would this fail during sim bc alliance isnt present?

Reply:
we might not sim this, idk

Reply:
I added that condition so that tests would work. during testing isPresent is false so it chooses blue speaker. if it is false the second part of the bool doesn't run (short circuit). it's safe.